### PR TITLE
Shapehelper reorg: add ShapeHelper for Einsum, OneHotEncoder, isNan, RandomLike, Reductions

### PR DESCRIPTION
**Pull request by AlexandreEichenberger**
Added shape inference ShapeHelper framework for the ops listed above. They used to compute shape inference without the framework.

In addition, removed the `#define` for Pooling ops, using template instead

### CODE REVIEW

File: src/Dialect/ONNX/ONNXOps/Math/ElementwiseBroadcast.cpp [hunk 1]
```diff
@@ -28,10 +28,10 @@ using namespace onnx_mlir;
 namespace {
 
 // Verify shape for numpy style broadcasting operators.
-template <class OP, class ADAPTOR>
+template <class OP_TYPE>
```
Comment (tungld) [inline src/Dialect/ONNX/ONNXOps/Math/ElementwiseBroadcast.cpp:31]: Had the same thought that `class ADAPTOR` was redundant when I was trying to use this function in NNPA.

File: src/Dialect/ONNX/ONNXOps/Math/ElementwiseBroadcast.cpp [hunk 2]
```diff
@@ -28,10 +28,10 @@ using namespace onnx_mlir;
...
 static LogicalResult verifyShapeForBroadcastingOps(
-    OP &op, Type elementType = nullptr) {
-  ADAPTOR operandAdaptor(op);
+    OP_TYPE &op, Type elementType = nullptr) {
```
Comment (tungld) [inline src/Dialect/ONNX/ONNXOps/Math/ElementwiseBroadcast.cpp:33]: Could we avoid template by using `Operation *op` instead, and move it to a common place, e.g. `ShapeHelper.cpp`? I want to use this function in NNPA code also. It seems `operandAdator` is used to get operands that can be obtained from `op->getOperands()`.

File: src/Dialect/ONNX/ONNXOps/Math/ElementwiseBroadcast.cpp (hunk 1 above)
Comment (AlexandreEichenberger) [reply src/Dialect/ONNX/ONNXOps/Math/ElementwiseBroadcast.cpp:31]: Yep, I saw this code from someone here, and decided to start using it more

File: src/Dialect/ONNX/ONNXOps/Math/ElementwiseBroadcast.cpp (hunk 2 above)
Comment (AlexandreEichenberger) [reply src/Dialect/ONNX/ONNXOps/Math/ElementwiseBroadcast.cpp]: did it in the latest version

**Review by tungld — APPROVED** [review]
LGTM.

**Pull request merged**